proxy/router/src/lib.rs

```diff
@@ -116,42 +134,55 @@ where T: Recognize,
     type Error = Error<T::Error, T::RouteError>;
     type Future = ResponseFuture<T>;
 
+    /// Always ready to serve.
+    ///
+    /// Graceful backpressure is **not** supported at this level, since each request may
+    /// be routed to different resources. Instead, requests should be issued and each
+    /// route should support a queue of requests.
+    ///
+    /// TODO Attempt to free capacity in the router.
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         Ok(().into())
     }
 
+    /// Routes the request through an underlying service.
+    ///
+    /// The response fails when the request cannot be routed.
     fn call(&mut self, request: Self::Request) -> Self::Future {
         let inner = &mut *self.inner.lock().expect("lock router cache");
 
-        let key = match inner.recognize.recognize(&request) {
-            Some(k) => k,
-            None => {
-                return ResponseFuture { state: State::NotRecognized };
-            }
-        };
+        match inner.recognize.recognize(&request) {
+            None => ResponseFuture::not_recognized(),
 
-        // Is the bound service for that key reusable? If `recognize`
-        // returned `SingleUse`, that indicates that the service may
-        // not be used to serve multiple requests.
-        if let Reuse::Reusable(ref k) = key {
-            if let Some(service) = inner.routes.get_mut(k) {
-                let response = service.call(request);
-                return ResponseFuture { state: State::Inner(response) };
+            Some(Reuse::SingleUse(key)) => {
+                // TODO Keep SingleUse services in the cache as well, so that their
+                // capacity is considered. To do this, we should move the Reuse logic into
+                // the returned service (and not the key).
```

Comment:
Hmm --- maybe I'm going down entirely the wrong path here, but WDYT of making the value type of the cache `Option<Service>` or `enum CacheEntry { Useabke<T>, SingleUse }` --- would there be any advantage for not storing the constructed single use services in the cache?

Reply:
as noted, this is all changing in a followup.